docs/proposals/scheduling/20241008-gpu-partition-api.md

```diff
@@ -0,0 +1,225 @@
+---
+title: GPU Partitioning APIs
+authors:
+- "@ZiMengSheng"
+reviewers:
+- "@hormes"
+- "@songtao98"
+- "@saintube"
+creation-date: 2024-10-08
+last-updated: 2023-10-08
+status: provisional
+---
+
+# GPU Partitioning APIs
+
+## Summary
+This proposal outlines an enhancement to the GPU scheduling capabilities of Koordinator, particularly focusing on NVIDIA GPUs operating under SharedNVSwitch mode. The primary objective is to introduce functionality that allows Pods to specifically request GPU partitions based on predefined configurations (Partitions).
+
+## Motivation
+In virtualized environments, when NVIDIA FabricManager operates in SharedNVSwitch mode, for security reasons, NVIDIA imposes certain requirements on the GPU configurations that can be allocated to a single VM, allowing only a few specific combinations of GPUs. NVIDIA refers to a combination of GPUs as a Partition and a table consisting of several such Partitions as a Partition Table.
+
+The scheduler in Koordinator is currently responsible for selection of GPUs for Pods. This PR expands upon the existing GPU scheduling capabilities of Koordinator, enabling it to recognize specific machine configurations and user requirements regarding GPU partitioning.
+
+### Goals
+- Provide the API for a Pod to request a specific GPU Partition.
+- Allow nodes to offer permitted Partitions and whether Pods are guaranteed to be allocated based on the partition.
+
+### Non-Goals/Future Work
+- Describe what the Partition Table looks like for a specific GPU model.
+
+## User Story
+
+Typically, the rules for GPU partitioning are determined by the specific GPU model or system configuration, and may also be influenced by the configuration of GPUs on each individual node. The scheduler does not have insight into the specifics of the hardware models or GPU types; instead, it relies on components at the node level to report these Partition Rules to the Device Custom Resource (CR) as follows:
+
+```yaml
+apiVersion: scheduling.koordinator.sh/v1alpha1
+kind: Device
+metadata:
+  annotations:
+    scheduling.koordinator.sh/gpu-partitions: |
+      {
+        "1": [
+            "NVLINK": {
+                {
+                  # Which GPUs are included
+                  "Minors": [
+                      0
+                  ],
+                  # GPU Interconnect Type
+                  "gpuLinkType": "NVLink",
+                  # GPU Interconnect Bandwidth
+                  "gpuLinkBandWidth": 400Gi
```

Comment:
gpuLinkBandWidth is ambiguous. The bandwidth supported by a topology is related to the collective communication operator. Consider changing it to ringAllreduceBandwidth?